src/kontrol/prove.py

```diff
@@ -656,23 +661,31 @@ def _final_cterm(
     failing: bool,
     is_test: bool = True,
     is_setup: bool = False,
+    hevm: bool = False,
 ) -> CTerm:
     final_term = _final_term(empty_config, program, is_test=is_test, is_setup=is_setup)
     dst_failed_post = KEVM.lookup(KVariable('CHEATCODE_STORAGE_FINAL'), Foundry.loc_FOUNDRY_FAILED())
-    foundry_success = Foundry.success(
-        KVariable('STATUSCODE_FINAL'),
-        dst_failed_post,
-        KVariable('ISREVERTEXPECTED_FINAL'),
-        KVariable('ISOPCODEEXPECTED_FINAL'),
-        KVariable('RECORDEVENT_FINAL'),
-        KVariable('ISEVENTEXPECTED_FINAL'),
-    )
     final_cterm = CTerm.from_kast(final_term)
-    if is_test:
-        if not failing:
-            return final_cterm.add_constraint(mlEqualsTrue(foundry_success))
-        else:
-            return final_cterm.add_constraint(mlEqualsTrue(notBool(foundry_success)))
+    if not hevm:
+        foundry_success = Foundry.success(
+            KVariable('STATUSCODE_FINAL'),
+            dst_failed_post,
+            KVariable('ISREVERTEXPECTED_FINAL'),
+            KVariable('ISOPCODEEXPECTED_FINAL'),
+            KVariable('RECORDEVENT_FINAL'),
+            KVariable('ISEVENTEXPECTED_FINAL'),
+        )
+        if is_test:
+            if not failing:
```

Comment:
Can you please check if `proveFail` is supported by `hevm` for declaring tests that are supposed to fail, similarly to Foundry? I saw that they used this convention in some of their materials, but maybe something changed since then.

Reply:
It seems it is not supported, I just tried with this simple example:
```solidity
function proveFail_simple2(uint x) public pure {
      require(x < 1);
      assert(x != 0);
```
And it fails with the command `hevm symbolic --code $BYTECODE --sig "proveFails_simple2(uint)"`:
```
Exploring contract
Simplifying expression
Explored contract (4 branches)
Checking for reachability of 1 potential property violation(s)

Discovered the following counterexamples:

Calldata:
  0x22a261690000000000000000000000000000000000000000000000000000000000000000

Storage:
  Addr SymAddr "miner": []
  Addr SymAddr "origin": []

Transaction Context:
  TxValue: 0x0
```

Reply:
Thanks for checking! Can you please also try `hevm test` and `halmos`? It looks like that might be our reference for what we're checking by default. ty!

Reply:
Yes, `hevm test` supports `proveFail`. But also, `hevm test` fails with `revert`, see the test `prove_revert` and `prove_expectRevert`.

Reply:
Thank you for confirming that! I just tried `proveFail` with Halmos, and it looks like it it's supported either. I think I'm slightly inclined to add support for `proveFail` to `hevm` success predicate checking (since `hevm test` supports this), but if don't find it reasonable, please feel free to not add it :) 
@RaoulSchaffranek can I ask for your opinion on that too? ty!

Reply:
Maybe we can also ask dxo what they think? I don't mind to add it :) 
What about the tests `prove_revert` and `prove_expectRevert`? Should we also make the results consistent with `hevm test`? According to the issue #312, it should not fail with `revert` statements, however `hevm test` fails. 

Also, just to confirm, should we check the hevm success predicate only for functions prefixed with `prove`?

Reply:
My suggestion is to:
- not report a test as failing on a failed `revert` or `expectRevert` (i.e., to not make it consistent with `hevm test` on `prove_revert` and `prove_expectRevert`)
- check the hevm success predicate on tests that are prefixed with `prove`, `check`, and `test`, as we do for Foundry tests (we did this to make our results compatible with those of Halmos and hevm, so I'd stick to that)
- support `proveFail` as a way to specify tests that are supposed to fail

But I don't have a strong opinion and would also check with @RaoulSchaffranek  and/or dxo, yes :)

Reply:
Implementing the `proveFail` is not that easy... For instance, the following test:
```solidity
function proveFail_require_assert(uint x) public pure {
      // Fails with `hevm symbolic --code $BYTECODE --sig "proveFail_require_assert(uint)"`
      // Passes with `hevm test`
      require(x < 1);
      assert(x != 0);
    }
```
Has two branches: 
1. `x < 1` -  hevm predicate returns `false` because of the user-defined assertion
2. `x >= 1` - hevm predicate returns `true`, although it ends with `EVMC_REVERT`

If we only make `proveFail` the symmetric to `prove`, as we do for Foundry, i.e.:
```python
if failing:
    return final_cterm.add_constraint(mlEqualsTrue(notBool hevm_success))
```
Then, the second branch will return `false`. We can work around this by implementing a different `hevm-fail` predicate that returns `true` if the function ends with `EVMC_REVERT`, but that would mean that the following tests would both pass, which would be inconsistent:
```solidity
function prove_revert() public pure {
      revert("Just reverts");
}

function proveFail_revert() public pure {
      revert("Just reverts");
}
```
Please let me know what you think. Or if you have a better solution :)

Reply:
Ok, I understood, thank you! I think `proveFail` tests should only fail if the current `hevm_success` predicate evaluates to `false`, e.g., if there's a failed user-defined or DSTest assertion. The problem here is that, in this case, the `require` statement plays a role of `vm.assume`, so, instead of branching on it, we should (ideally, in this `hevm`-like execution mode) just add it to the list of path conditions.

Reply:
I agree that's problematic, so maybe we can leave it for another PR. My only worry is that there'll be some discrepancy between our Foundry and hevm modes. Do you think it makes sense to add an info message for the user if `--hevm` is on and an `xFail` test is being analyzed?

Reply:
I agree that would be the best implementation. I'll try to do that in this PR, it shouldn't be that hard. In case it is too complex, I'll go with the second option - add an info message for the `xFail` - and leave that for another PR.

Reply:
> The problem here is that, in this case, the require statement plays a role of vm.assume, so, instead of branching on it, we should (ideally, in this hevm-like execution mode) just add it to the list of path conditions.

I'm strongly in favor of this implementation. It allows us to write meaningful specifications in pure Solidity without the forge-std library - which can be very handy.

From the discussion it appears that `hevm` uses different success predicates for `hevm test` and `hevm symbolic`. And we're trying to simulate `hevm test`, is my understanding correct?

Reply:
> From the discussion it appears that `hevm` uses different success predicates for `hevm test` and `hevm symbolic`. And we're trying to simulate `hevm test`, is my understanding correct?

It looks like we're trying to simulate some middle ground between `hevm test` and `halmos` :). DSTest assert violations are supported by `hevm test`, so here we follow it, yes; but @lisandrasilva also mentioned that `hevm test` supports, for example, `expectRevert` statements, which we should ignore in the "user-defined tests and dstest asserts" success predicate (if I understand correctly).